### ISSUE TITLE
a timeSeries data reading problem

### ISSUE THREAD

**Issue by cqiaoYc** [deeplearning4j/deeplearning4j#3988]
#### Issue Description
 the problem using SequenceRecordReaderDataSetIterator.class 
program execution is very fast, training monitoring UI displays nothing,  the predict valuses all are class-0. Tracing the running, finds that the feature data is organized in columns.           
   It's SingleTimestep prodiction. I save time series data in the CSV file, each sequence of a file, each line of a time step. Each row has 4 feature columns and a tag column, such as:           
   3.17,-1.58,0.00,-0.92,3
   0.40,1.19,-1.58,-0.61,2
   1.19,-0.79,0.00,-0.87,2
   0.79,0.40,-0.40,0.17,2
   0.79,-0.40,-0.40,0.29,3
Data read and training, test code as follows: 
```
   private DataSetIterator getDataSetIter(String fileName, int maxFileNumb) {
       try {
           SequenceRecordReader reader = new CSVSequenceRecordReader(0, ",");
           reader.initialize(new NumberedFileInputSplit(normalizedDataSaveDir + fileName + "_%d.csv", 0, maxFileNumb));
           return new SequenceRecordReaderDataSetIterator(reader, batchSize, 5, 4); //numClassLabel=5, labelIndex=4
       } catch (IOException | InterruptedException ex) {
           Logger.getLogger(PriceChangePrecastModelLstmImpl2.class.getName()).log(Level.SEVERE, null, ex);
       }
       return null;
   }
    public MultiLayerNetwork train() {
       this.trainDataSetIter = this.getDataSetIter("train", 433);
       this.testDataSetIter = this.getDataSetIter("test", 185);
       MultiLayerNetwork model = new MultiLayerNetwork(getNetConf());
       model.init();//     
       uiMonitor(model);
 
       int nEpochs = 1;
//        int batchSize = 64;
       for (int i = 0; i < nEpochs; i++) {
           trainDataSetIter.reset();            
           model.fit(trainDataSetIter);
       }
       return model;
   }
 
   public void test(MultiLayerNetwork model) {
       test(model, this.trainDataSetIter, "train");
       test(model, this.testDataSetIter, "test");
   }
 
   public void test(MultiLayerNetwork model, DataSetIterator dataSetIter, String description) {
       dataSetIter.reset();
       Evaluation eval = model.evaluate(dataSetIter, generateLables());
 
       System.out.println(description + " stats: " + eval.stats());
       System.out.println(description + " confusionMatrix: ");
       System.out.println(eval.confusionToString());
   }
```
#### Version Information
* Deeplearning4j version 0.9.1

[cqiao_trainData.zip](https://github.com/deeplearning4j/deeplearning4j/files/1266523/cqiao_trainData.zip)

**Comment by cqiaoYc** [comment]
If  the test method is changed to:
```
while (dataSetIter.hasNext()) {
            if (trainDataSetIter.isNewSeries()) {
                 model.rnnClearPreviousState();
            }
            DataSet dataSet = dataSetIter.next();   //note here!
            INDArray labels=dataSet.getLabels();
            INDArray predicts = model.rnnTimeStep(dataSet.getFeatures());
            eval.evalTimeSeries(labels, predicts);
        }
       ...
     ```
An exception is thrown:
Exception in thread "main" org.deeplearning4j.exception.DL4JInvalidInputException: Previous activations (stored state) number of examples = 64 but input array number of examples = 29. Possible cause: using rnnTimeStep() without calling rnnClearPreviousState() between different sequences?
	at org.deeplearning4j.nn.layers.recurrent.LSTMHelpers.activateHelper(LSTMHelpers.java:178)
	at org.deeplearning4j.nn.layers.recurrent.GravesLSTM.activateHelper(GravesLSTM.java:159)
	at org.deeplearning4j.nn.layers.recurrent.GravesLSTM.rnnTimeStep(GravesLSTM.java:243)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.rnnTimeStep(MultiLayerNetwork.java:2651)
	at com.cq.mltrademodels.PriceChangePrecastModelLstmImpl.test(PriceChangePrecastModelLstmImpl.java:561)

**Comment by cqiaoYc** [comment]
The structures of features in memory are different using my iterator and using SequenceRecordReaderDataSetIterator.class. The features in memory  using my iterator is like:
features=(org.nd4j.linalg.jcublas.JCublasNDArray)[[3.44,-1.72,0.00,3.16],
[0.43,1.29,-1.72,1.70],
[1.29,-0.86,0.00,0.49],
[0.86,0.43,-0.43,0.65],...
The features in memory  using SequenceRecordReaderDataSetIterator.class is like:
features=(org.nd4j.linalg.jcublas.JCublasNDArray)[[[3.17,0.40,1.19,0.79,0.79,0.40,0.40,0,00,0.79,0.40,0.00,...

the stats using my iterator:
# of classes:    0
 Accuracy:        0.0000
 Precision:       NaN
 Recall:          NaN
 F1 Score:        NaN

**Comment by AlexDBlack** [comment]
I'm not going to go through and debug your custom iterator for you. You'll have to write your own unit tests for that to check it is correct.

> I save time series data in the CSV file, each sequence of a file, each line of a time step.

If you are using CSV for your data, why can't you use CsvSequenceRecordReader? That's properly tested and works...

**Comment by cqiaoYc** [comment]
I used CSVSequenceRecordReader:
```
private DataSetIterator getDataSetIter(String fileName, int maxFileNumb) {
       try {
           **SequenceRecordReader reader = new CSVSequenceRecordReader(0, ",")**;  
           reader.initialize(new NumberedFileInputSplit(normalizedDataSaveDir + fileName + "_%d.csv", 0, maxFileNumb));
           return new SequenceRecordReaderDataSetIterator(reader, batchSize, 5, 4); //numClassLabel=5, labelIndex=4
       } catch (IOException | InterruptedException ex) {
           Logger.getLogger(PriceChangePrecastModelLstmImpl2.class.getName()).log(Level.SEVERE, null, ex);
       }
       return null;
   }

``

**Comment by cqiaoYc** [comment]
I tranced the running, found the all input of outputLayer were almost 0s. my config:
```
        int iterations = 1;
        long seed = 200;
        double learnRate = 5e-3f;
        MultiLayerConfiguration conf = new NeuralNetConfiguration.Builder()
                .seed(seed)
                .iterations(iterations)
                .weightInit(WeightInit.XAVIER)
                .optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT)
                .learningRate(learnRate)
                .updater(new RmsProp(0.95))
                .regularization(true).l2(1e-3)
                .gradientNormalization(GradientNormalization.RenormalizeL2PerLayer)
                .list()
                .layer(0, new GravesLSTM.Builder().activation(Activation.SOFTSIGN).nIn(this.featureCount).nOut(this.featureCount * 4).learningRate(1e-2f).build())
                .layer(1, new GravesLSTM.Builder().activation(Activation.SOFTSIGN).nIn(this.featureCount * 4).nOut(this.featureCount * 3).learningRate(5e-3f).build())
                .layer(2, new GravesLSTM.Builder().activation(Activation.SOFTSIGN).nIn(this.featureCount * 3).nOut(this.featureCount * 2).learningRate(1e-2f).build())
                .layer(3, new RnnOutputLayer.Builder().lossFunction(new LossMCXENT(getClassWeights())).weightInit(WeightInit.XAVIER)
                        .activation(Activation.SOFTMAX).nIn(featureCount * 2).nOut(this.levelCount*2+1).build())                
//                 .backpropType(BackpropType.TruncatedBPTT).tBPTTForwardLength(20).tBPTTBackwardLength(20)
                .pretrain(false)
                .backprop(true)
                .build();
        return conf;
```

**Comment by cqiaoYc** [comment]
@AlexDBlack After tracking code execution, make sure my iterator is ok. The eval.stats() is incorrect because it is used: new valuation () .

**Comment by eraly** [comment]
There are many reasons the net won't learn. 

```
Accuracy is low or network isn't learning? Start by reading http://deeplearning4j.org/troubleshootingneuralnets and http://deeplearning4j.org/visualization
```
